dramatiq/composition.py

```diff
@@ -224,7 +242,51 @@ def run(self, *, delay=None):
           delay(int): The minimum amount of time, in milliseconds,
             each message in the group should be delayed by.
         """
-        for child in self.children:
+        if self.completion_callbacks:
+            from .middleware.group_callbacks import GROUP_CALLBACK_BARRIER_TTL, GroupCallbacks
+
+            rate_limiter_backend = None
+            for middleware in self.broker.middleware:
+                if isinstance(middleware, GroupCallbacks):
+                    rate_limiter_backend = middleware.rate_limiter_backend
+                    break
+            else:
+                raise RuntimeError(
+                    "GroupCallbacks middleware not found! Did you forget "
+                    "to set it up? It is required if you want to use "
+                    "group callbacks."
+                )
+
+            # Generate a new completion uuid on every run so that if a
+            # group is re-run, the barriers are all separate.
+            # Re-using a barrier's name is an unsafe operation.
+            completion_uuid = str(uuid4())
+            completion_barrier = Barrier(rate_limiter_backend, completion_uuid, ttl=GROUP_CALLBACK_BARRIER_TTL)
+            completion_barrier.create(len(self.children))
```

Comment:
While writing #144, I determined that while the first barrier in a chain of groups should be created immediately, I think that callback groups need to have their barriers created by the middleware. I recognize that your implementation doesn't, yet, support groups as callbacks, but I do think that you'll want to do that.

This is what I created the `groups_starts` option for. And the matching `GroupStart` data class, but that's an implementation detail, it can be whatever form it needs, but it needs to allow for signaling (1) which group to start, and (2) how many tasks are in the group that will start.

In order to support directly nested groups (still raises a `NotImplementedError` in your current implementation, but written in such a way to suggest to me that you may be intending to), you'll need to allow for starting multiple groups at once.

Given the other options you're adding in this PR, I'll suggest that this should be called `group_completion_starts`, and have a shape like this:

```python
message.options['group_completion_starts'] = [
    {'uuid': 'the-first-uuid', 'count': 25},
    {'uuid': 'the-second-uuid', 'count': 50},
]
```